### PR TITLE
Scripts and processes - Script output

### PR DESCRIPTION
**Pull request by jonas-atmire**
## References
_Add references/links to any related issues or PRs. These may include:_
* Related to [REST Contract](https://github.com/DSpace/Rest7Contract/blob/main/scripts-endpoint.md) -> Pertaining to the "output" link -> "/api/system/processes/5/output"
* Fixes [GitHub issue](https://github.com/DSpace/DSpace/pull/2820#discussion_r459276967), the output of the script can afterwards be used for additional testing etc
* Related to discussion from [PR](https://github.com/DSpace/DSpace/pull/2905)

## Description
For both the Angular team, as well as for the writing of certain REST tests, we want to be able to see the output of a script that has been started from the scripts and processes.

## Instructions for Reviewers

This code stores the output in the assetstore (attached to the process), but only after the script completed or failed
In the meantime, while the script is running, the output is temporarily kept as a temporary file.
The output would be made available as a straight bitstream so that angular can parse it to the output format.

**Please note** that there are still some hibernate related problems that are currently failing the tests. (Work in progress)
The problem with hibernate here is most likely due to the order of deletion of the bitstreams.
Another weird thing that is going on that, when running the test manually, the tests do not seem to fail consistently 'similar' all the time (So the tests are still failing, but in a different way)

## Checklist
_This checklist provides a reminder of what we are going to look for when reviewing your PR. You need not complete this checklist prior to creating your PR (draft PRs are always welcome). If you are unsure about an item in the checklist, don't hesitate to ask. We're here to help!_

- [x] My PR is small in size (e.g. less than 1,000 lines of code, not including comments & integration tests). Exceptions may be made if previously agreed upon.
- [x] My PR passes Checkstyle validation based on the [Code Style Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Style+Guide).
- [x] My PR includes Javadoc for _all new (or modified) public methods and classes_. It also includes Javadoc for large or complex private methods.
- [x] My PR passes all tests and includes new/updated Unit or Integration Tests based on the [Code Testing Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Testing+Guide).
- [x] If my PR includes new, third-party dependencies (in any `pom.xml`), I've made sure their licenses align with the [DSpace BSD License](https://github.com/DSpace/DSpace/blob/main/LICENSE) based on the [Licensing of Contributions](https://wiki.lyrasis.org/display/DSPACE/Code+Contribution+Guidelines#CodeContributionGuidelines-LicensingofContributions) documentation.
- [x] If my PR modifies the REST API, I've linked to the REST Contract page (or open PR) related to this change.

### CODE REVIEW

**Comment by tdonohue** [comment]
@jonas-atmire : This is currently failing cause of checkstyle violations. In addition, it looks like it introduces two new warnings in LGTM (which would be nice to resolve), see https://github.com/DSpace/DSpace/pull/2934#issuecomment-677634271

**Comment by tdonohue** [comment]
@jonas-atmire : This PR looks to be failing it's own tests. 

Additionally, I believe LGTM is pointing out a [flaw here](https://github.com/DSpace/DSpace/pull/2934#issuecomment-678135629) in the fact that you are (accidentally) using the `scriptName` param passed in on the URL path to determine the path of the log files. There's a risk here that would allow users to possibly change the path of the log files (if somehow they figured out a way to pass in something like `../../../script` on the URL).   

All that said, I believe you could fix that potential bug easily with a one-line update to `ScriptRestRepository` on these lines: https://github.com/DSpace/DSpace/blob/main/dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/ScriptRestRepository.java#L104-L105

I think those lines should be updated to NOT use `scriptName` (which is the passed in param), but instead get the name of the script from the `scriptToExecute` object like this:
```
RestDSpaceRunnableHandler restDSpaceRunnableHandler = new RestDSpaceRunnableHandler(
            context.getCurrentUser(), scriptToExecute.getName(), dSpaceCommandLineParameters);
```
That should remove the warning from LGTM, as the passed in user param will only be used to locate the `scriptToExecute`, and then after that the `scriptToExecute.getName()` will end up being used to decide the name of the log files you are creating in this PR.

Once these updates are made, let me know & I'd be glad to review & test this PR.

**Comment by jonas-atmire** [comment]
@tdonohue 
You are correct in mentioning that the tests still fail, I had mentioned this in the description of the PR, but could have made it more clear. I'd opted to already create the PR rather than spending an unknown amount of time trying to investigate the tests issue, in the hope that someone might have had a similar problem in the past. (That being said: I'll have the last LGTM problem fixed as well, I'd overlooked this sadly)

**Quick recap of tests problem**
The weird thing here is that the tests do not seem to be failing consistently.
Some of the times, it fails because of the process number not matching what is being expected (And it also incrementing in the meanwhile)
```
Expected: a string containing "INFO mock-script - 1 @ The script has started" 
but: was "2020-08-21 09:59:23.431 INFO mock-script - 42
....
```
☝️  This leads me to believe that the processes might not be properly cleared up after the test run has finished.

Other times it seems to fail in some hibernate related way where it tries to delete a bitstream that is marked as 'obsolete'
This looks something more 'timing' related, especially in combination with the first error also sometimes showing up (As in: the hibernate states are not always up to date with the code being run, or vice versa)

**Comment by tdonohue** [comment]
@jonas-atmire : When tests randomly fail in JUnit, 95% (or more) of the time it's because one (or more) tests are not cleaning up after themselves, or somehow written incorrectly.  The way this test is behaving seems odd to me if you look closely at how it is failing.  It almost seems to imply something is wrong with the code.

The failing portion of the `postProcessAndVerifyOutput()` seems to start with the section on line 335, where you are simply getting the Bitstream associated with the current Process: https://github.com/DSpace/DSpace/pull/2934/files#diff-c33ee454d5001ee173374f191a53bfa3R335

Strangely, though, that processService.getBitstream() call seems like it is returning logs for a DIFFERENT process, as the test is essentially saying you are calling getBitstream() for Process with ID=1, but getting a Bitstream which has the logs for Process 42 instead.  That doesn't seem like anything related to Hibernate...and honestly, it might not even be a test cleanup problem.  It almost sounds like, in some scenarios either (a) the getBitstream() call is returning the *wrong Bitstream ID*, or (b) the content logs within that Bitstream are logging the *wrong Process ID* (i.e. maybe some code is not multithreaded?), or (c) Processes are somehow not writing their logs to the correct Bitstream (also could be a multithreaded issue).

I'm honestly not sure what is going on here exactly, but the type of failure does not seem (to me) like it could have anything to do with Hibernate.  It almost seems like a synchronization flaw (some code is expecting to be single threaded, but is obviously not).  But, I could be wrong.

As a sidenote, this PR has merge conflicts.  So, it's hard for me to actually pull down the code and try tests myself.  Everything above is simply from my reading of the code & what the test error seems to say (to me).  If you cannot find the issue, I might be able to find time next week to try things out locally (after you fix the conflicts).  Thanks

**Comment by jonas-atmire** [comment]
@tdonohue 
We were indeed in the process of both investigating the problem as well as fixing the merge conflict.

The problem with the test execution was mostly because normally the tests should be using the SyncTaskExecutor to ensure that the tasks are fully completed before proceeding. But the Threaded version seemed to be used previously.
This resulted in the previously mentioned problems.
The tests are now properly coming through and are passing correctly.

I've removed the 'work in progress' label and updated the 'test' related checkbox of the description accordingly, this is now ready for more extensive, functional testing as well.

**Review by tdonohue — CHANGES_REQUESTED** [review]
@jonas-atmire : Code is looking good overall, though I've added some inline comments below on areas that could use more JavaDocs, comments or explanation.  Most should be easy to resolve.

I also gave this some light testing & it looks like it's mostly working on it's own.  I'm having issues getting it to work *with* the Angular PR (https://github.com/DSpace/dspace-angular/pull/827), but I *think* the issues might be on the Angular side.

File: dspace/config/spring/api/scripts.xml
Comment (jonas-atmire) [reply dspace/config/spring/api/scripts.xml]: **Could you add a comment to explain why this is set to 5?**

There is no particular reason that this is set to 5, it’s just a default to not allow too many threads to be working at the same time and potentially cause memory issues or other performance issues.  This was [previously also set to 5](https://github.com/DSpace/DSpace/blob/05231676253d397262569f1f5386b6b297547b22/dspace/config/spring/rest/scripts.xml#L6), so we kept this as is

**I’m also curious why the production version of `dspaceRunnableThreadExecutor` is of type `ThreadPoolTaskExecutor`, while all the tests use `SyncTaskExecutor`. Shouldn’t the tests use the same type of TaskExecutor as the main code? If not, I think we should add an inline comment to describe the reason for the difference (it’s not obvious to me).**

The DSpaceRunnableThreadExecutor is responsible for executing the Process. This Process will either run synchronously or asynchronously.

In the **production code**, it’s desirable that this happens asynchronously because we don’t want the user to wait for his/her REST response until a Process is done with the execution, it’ll return almost immediately and give a “running” or “scheduled” if it’s not done yet, but it’ll continue to run in a different thread and the Process db object will be updated when this background thread is done running.
This way the user won’t ever have to wait ages for the REST reply even when executing large scripts. We don’t want the user to wait or any front-ends to be blocked whilst waiting for a response on the call when it could just be ran in the background and update the db when it’s finished executing.

In the **test code** however, we want to actually test things after a REST call is made. This means that we can’t have processes making alterations to the DB or be stuck in execution whilst the assertions are being tested in the tests. The result that’s coming from the REST api has to be consistent and represent a state in which the Process is fully finished executing, whether it’s completed or failed. This is necessary because otherwise our tests aren’t going to be the same every time they’re ran and they’ll cause failures and inconsistencies.

Implementing this SyncThreadExecutor ensures that any Processes started by the DSpaceRunnableThreadExecutor will have to fully run and be fully done before the REST response is sent. This means that we can check for completed or failed instead of sometimes running into “running” status when the Process would still be executing. We’re simply forcing the Process to fully finish its execution before we construct the REST response so that it’s always consistent and non-variable.

If you remember from a while back, [I mentioned that the tests were failing inconsistently](https://github.com/DSpace/DSpace/pull/2934#issuecomment-682358229), this was mostly because of the executor that was used.
Because of the asynchronous run that was being done at that point in time, the tests did not correctly await the responses/checked the assertions as was described above.
We did removed an unneeded dspaceRunnableThreadExecutor in the api scripts.xml. This is a bean that’s only needed in the server module, so it should be in the REST package only if I’m not horribly mistaken

**Review by tdonohue — APPROVED** [review]
👍 I've retested & re-reviewed.  This looks good to me now. Thanks for the answers as well @jonas-atmire . 

I tested this with the UI PR (https://github.com/DSpace/dspace-angular/pull/827) and found bugs at the UI layer (see [my latest review](https://github.com/DSpace/dspace-angular/pull/827#pullrequestreview-484507417) there), but they don't seem to be caused by this backend PR.

**Review by benbosman — APPROVED** [review]
I've reviewed the code and tested it.
I didn't see any code issues, seems Tim already found them.

It successfully creates the log file, deletes it on the end of the script, and I can access it once completed.
If I shut down tomcat during the execution of the script, the log file remains available on disk

**Comment by tdonohue** [comment]
Merging, as this is at +2.

**Pull request merged**
